core-kotlin-modules/core-kotlin-concurrency-3/src/test/java/com/baeldung/singleRxJavaToCoroutineDeferred/SingleRxJavaToCoroutineDeferredUnitTest.kt

```diff
@@ -0,0 +1,216 @@
+        deferred.assertResultsTrue() // assertion test
+    }
+
+    // using async with extension
+    private suspend fun <T : Any> Single<T>.toDeferredAsync(): Deferred<T> =
+        coroutineScope { async { this@toDeferredAsync.blockingGet() } }
+
+
+    @Test
+    fun `test using async with extension`() = runBlocking {
+        val deferredExt = getFilteredProducts().toDeferredAsync()
+        deferredExt.assertResultsTrue()
+    }
+
+    // using CoroutineScope(context).async
+    private fun <T : Any> Single<T>.toDeferredWithContext(context: CoroutineContext): Deferred<T> =
+        CoroutineScope(context).async { this@toDeferredWithContext.blockingGet() }
+
+    @Test
+    fun `test using CoroutineScope with context and async`(): Unit = runBlocking {
+        val deferred = getFilteredProducts().toDeferredWithContext(Dispatchers.IO)
+        deferred.assertResultsTrue()
+    }
+
+    // using CompletableDeferred & subscribe
+    @Test
+    fun `using CompletableDeferred & subscribe`() = runBlocking {
+        val deferred = CompletableDeferred<List<Product>>()
+        getFilteredProducts()
+            .subscribe({ products ->
+                    deferred.complete(products)
+                },{ error ->
+                    deferred.completeExceptionally(error)
+                }
+            )
+        deferred.assertResultsTrue()
+    }
+
+    // using CompletableDeferred extension
+    private fun <T : Any> Single<T>.toCompletableDeferred(): CompletableDeferred<T> {
+        val completableDeferred = CompletableDeferred<T>()
+        this.subscribe({ completableDeferred.complete(it) }, { completableDeferred.completeExceptionally(it) })
+        return completableDeferred
+    }
+
+    // using CompletableDeferred with custom callback
+    private fun <T : Any> Single<T>.toCompletableDeferred(
+        onSuccess: (CompletableDeferred<T>, T) -> Unit, onError: (CompletableDeferred<T>, Throwable) -> Unit
+    ): CompletableDeferred<T> {
+        val completableDeferred = CompletableDeferred<T>()
+        this.subscribe({ result ->
+            completableDeferred.complete(result)
+            onSuccess(completableDeferred, result)
+        }, { error ->
+            completableDeferred.completeExceptionally(error)
+            onError(completableDeferred, error)
+        })
+        return completableDeferred
+    }
+
+    @Test
+    fun `test using CompletableDeferred extension`() = runBlocking {
+        val deferred = getFilteredProducts().toCompletableDeferred()
+        deferred.assertResultsTrue()
+    }
+
+    @Test
+    fun `test using CompletableDeferred with callback`(): Unit = runBlocking {
+        getFilteredProducts().toCompletableDeferred(onSuccess = { deferredResult, _ ->
+            launch { deferredResult.assertResultsTrue() }
+        }, onError = { _, error ->
+            println("Error: ${error.message}")
```

Comment:
Don't forget the guideline that we can't use println and should use SLF4J for logging